packages/container/libraries/strongly-typed/src/container.ts

```diff
@@ -0,0 +1,197 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable @typescript-eslint/no-unnecessary-type-parameters */
+import type { interfaces } from 'inversify';
+
+type IfAny<T, TYes, TNo> = 0 extends 1 & T ? TYes : TNo;
+
+type BindingMapProperty = string | symbol;
+type BindingMap = Record<BindingMapProperty, any>;
+type BindingMapKey<T extends BindingMap> = keyof T & BindingMapProperty;
+type MappedServiceIdentifier<T extends BindingMap> = IfAny<
+  T,
+  interfaces.ServiceIdentifier,
+  BindingMapKey<T>
+>;
+type ContainerBinding<
+  T extends BindingMap,
+  TKey extends MappedServiceIdentifier<T> = any,
+> = TKey extends keyof T
+  ? T[TKey]
+  : TKey extends interfaces.Newable<infer C>
+    ? C
+    : TKey extends interfaces.Abstract<infer D>
+      ? D
+      : never;
+
+type First<T extends any[]> = T extends [infer TFirst, ...any[]]
+  ? TFirst
+  : never;
+
+type AllButFirst<T extends any[]> = T extends [any, ...infer TRest]
+  ? TRest
+  : never;
+
+interface ContainerOverrides<
+  T extends BindingMap = any,
+  TAncestors extends BindingMap[] = any[],
+> {
+  parent: ContainerOverrides<First<TAncestors>, AllButFirst<TAncestors>> | null;
+  bind: Bind<T>;
+  get: <
+    TBound extends ContainerBinding<T, TKey>,
+    TKey extends MappedServiceIdentifier<T> = any,
+  >(
+    serviceIdentifier: TKey,
+  ) => TBound;
+  getNamed: <
+    TBound extends ContainerBinding<T, TKey>,
+    TKey extends MappedServiceIdentifier<T> = any,
+  >(
+    serviceIdentifier: TKey,
+    named: PropertyKey,
+  ) => TBound;
+  getTagged: <
+    TBound extends ContainerBinding<T, TKey>,
+    TKey extends MappedServiceIdentifier<T> = any,
+  >(
+    serviceIdentifier: TKey,
+    key: PropertyKey,
+    value: unknown,
+  ) => TBound;
+  getAll: <
+    TBound extends ContainerBinding<T, TKey>,
+    TKey extends MappedServiceIdentifier<T> = any,
+  >(
+    serviceIdentifier: TKey,
+  ) => TBound[];
+  getAllTagged: <
+    TBound extends ContainerBinding<T, TKey>,
+    TKey extends MappedServiceIdentifier<T> = any,
+  >(
+    serviceIdentifier: TKey,
+    key: PropertyKey,
+    value: unknown,
+  ) => TBound[];
+  getAllNamed: <
+    TBound extends ContainerBinding<T, TKey>,
+    TKey extends MappedServiceIdentifier<T> = any,
+  >(
+    serviceIdentifier: TKey,
+    named: PropertyKey,
+  ) => TBound[];
+  getAsync: <
```

Comment:
Need to change the types for `get()` to **not** allow any mapped values that return a `Promise`, and **only** allow these bindings to be accessed by the async methods